File: srtf.py
#!/usr/bin/env python3
"""SRTF (Shortest Remaining Time First) - Preemptive SJF"""

def srtf(processes):
    """processes: list of (pid, arrival_time, burst_time)"""
    procs = sorted(processes, key=lambda p: p[1])
    n = len(procs)
    remaining_time = {p[0]: p[2] for p in procs}
    completion_time = {}
    time = 0
    completed = 0
    
    # find max time needed
    max_time = max(p[1] for p in procs) + sum(p[2] for p in procs)
    
    while completed < n and time < max_time:
        # find process with shortest remaining time that has arrived
        available = [p for p in procs if p[1] <= time and remaining_time[p[0]] > 0]
        
        if available:
            # pick shortest remaining time
            current = min(available, key=lambda p: (remaining_time[p[0]], p[1]))
            pid = current[0]
            
            # execute for 1 time unit
            remaining_time[pid] -= 1
            time += 1
            
            # check if completed
            if remaining_time[pid] == 0:
                completion_time[pid] = time
                completed += 1
        else:
            time += 1
    
    # build result
    result = []
    for pid, arrival, burst in procs:
        ct = completion_time[pid]
        tat = ct - arrival
        wt = tat - burst
        result.append((pid, arrival, burst, ct, tat, wt))
    
    return result


def print_table(result):
    print("{:>4} {:>4} {:>4} {:>4} {:>5} {:>5}".format("PID", "AT", "BT", "CT", "TAT", "WT"))
    for row in result:
        print("{:>4} {:>4} {:>4} {:>4} {:>5} {:>5}".format(*row))
    n = len(result)
    avg_tat = sum(r[4] for r in result) / n
    avg_wt = sum(r[5] for r in result) / n
    print(f"\nAverage Turnaround Time = {avg_tat:.2f}")
    print(f"Average Waiting Time    = {avg_wt:.2f}")


if __name__ == '__main__':
    processes = [
        (1, 0, 8),
        (2, 1, 4),
        (3, 2, 2),
        (4, 3, 1),
    ]
    res = srtf(processes)
    print("SRTF (Shortest Remaining Time First):\n")
    print_table(res)

File: srtf.js
#!/usr/bin/env node
/**
 * SRTF (Shortest Remaining Time First) - Preemptive SJF
 */

function srtf(processes) {
    const procs = [...processes].sort((a, b) => a.arrival - b.arrival);
    const n = procs.length;
    const remainingTime = {};
    const completionTime = {};
    
    procs.forEach(p => {
        remainingTime[p.pid] = p.burst;
    });
    
    let time = 0;
    let completed = 0;
    const maxTime = Math.max(...procs.map(p => p.arrival)) + procs.reduce((sum, p) => sum + p.burst, 0);
    
    while (completed < n && time < maxTime) {
        // find process with shortest remaining time that has arrived
        const available = procs.filter(p => p.arrival <= time && remainingTime[p.pid] > 0);
        
        if (available.length > 0) {
            // pick shortest remaining time
            available.sort((a, b) => remainingTime[a.pid] - remainingTime[b.pid] || a.arrival - b.arrival);
            const current = available[0];
            
            // execute for 1 time unit
            remainingTime[current.pid]--;
            time++;
            
            // check if completed
            if (remainingTime[current.pid] === 0) {
                completionTime[current.pid] = time;
                completed++;
            }
        } else {
            time++;
        }
    }
    
    // build result
    const result = procs.map(p => {
        const ct = completionTime[p.pid];
        const tat = ct - p.arrival;
        const wt = tat - p.burst;
        return {
            pid: p.pid,
            arrival: p.arrival,
            burst: p.burst,
            completion: ct,
            turnaround: tat,
            waiting: wt
        };
    });
    
    return result;
}

function printTable(result) {
    console.log("PID".padStart(4), "AT".padStart(4), "BT".padStart(4), 
                "CT".padStart(4), "TAT".padStart(5), "WT".padStart(5));
    
    for (const row of result) {
        console.log(
            String(row.pid).padStart(4),
            String(row.arrival).padStart(4),
            String(row.burst).padStart(4),
            String(row.completion).padStart(4),
            String(row.turnaround).padStart(5),
            String(row.waiting).padStart(5)
        );
    }
    
    const n = result.length;
    const avgTAT = result.reduce((sum, r) => sum + r.turnaround, 0) / n;
    const avgWT = result.reduce((sum, r) => sum + r.waiting, 0) / n;
    
    console.log(`\nAverage Turnaround Time = ${avgTAT.toFixed(2)}`);
    console.log(`Average Waiting Time    = ${avgWT.toFixed(2)}`);
}

// Main execution
if (require.main === module) {
    const processes = [
        { pid: 1, arrival: 0, burst: 8 },
        { pid: 2, arrival: 1, burst: 4 },
        { pid: 3, arrival: 2, burst: 2 },
        { pid: 4, arrival: 3, burst: 1 }
    ];
    
    const result = srtf(processes);
    console.log("SRTF (Shortest Remaining Time First):\n");
    printTable(result);
}

module.exports = { srtf, printTable };
